### ISSUE TITLE
File the bug with jigsaw-dev about leaf modules and `--add-reads`

### ISSUE THREAD

**Issue by eljobe** [gradle-guides/building-java-9-modules#4]
We really should have to explain that the newly patched module is going to read the junit automatic module, but we don't have to for leaf projects like the "actors" project. The jigsaw-dev mailing list needs to know.

**Comment by eljobe** [comment]
Holy Cow! This isn't actually a "bug." It is sort of a "not working" as designed situation. Here's the crazy thing that is going on:

**Symptom**
To run the tests for the module `org.gradle.actors` we do not have to specify the `--add-modules junit` or `--add-reads org.gradle.actors=junit` to the java compiler when compiling the tests. Neither do we need to add `--add-reads org.gradle.actors=junit` to the runtime invocation of the java command when running the tests.

However, all of these arguments **are** needed for the `org.gradle.fairy.tale.formula` module which requires `org.gradle.actors`.  Why?

**Answer**
It's because of the third line in the `module-info.java` file for `org.gradle.actors`.

```java
module org.gradle.actors {
    exports org.gradle.actors;
    requires guava;  // <-- That also implies you read junit.
}
```

Now, if you haven't already carefully read [section 3.3 of "The State of the Module System"](http://openjdk.java.net/projects/jigsaw/spec/sotms/#automatic-modules), you're probably surprised that requiring the `guava` module also means you can view and read the `junit` module.

But, once you've read and internalized the horror that is that section of the document, you'll understand why it works that way. Here's the brief summary:

It is because `guava` is an automatic module on the module path. As soon as you have required access to one automatic module, then you can read any class from any package in any automatic module on the module path. Why? Because, without processing all of the class files in the module path, you cannot know that the code in the automatic `guava` module doesn't read some class from the `junit` module, or that it doesn't transitively need to expose the classes defined in that module because they are part of the `guava` modules exported api.

In summary, the part of the new module system which handles automatic modules is designed to be very permissive. You'll have fewer surprises if you can avoid automatic modules all-together. But, once you depend on just one of them explicitly, you also gain access to all of them implicitly.
